File: sdks/typescript/src/agent/identity.ts
/**
 * Who and where a connected agent is: its environment, its instance identity
 * and the endpoint it connects to. Every read of the machine is defensive, so
 * a locked-down sandbox with no hostname or no passwd entry still connects.
 */

import * as os from "node:os";
import { randomUUID } from "node:crypto";
import { LANGWATCH_SDK_VERSION } from "../internal/constants";
import { resolveEndpoint } from "../internal/endpoint";
import type { RegisterInstance, RegisterSdk } from "./protocol";

export const DEFAULT_ENVIRONMENT = "development";
const ENVIRONMENT_MAX_LENGTH = 32;

/** The environment variables read in order after the explicit option. */
const ENVIRONMENT_VARIABLES = [
  "LANGWATCH_AGENT_ENVIRONMENT",
  "APP_ENV",
  "ENVIRONMENT",
  "NODE_ENV",
] as const;

const isSet = (value: string | undefined): value is string =>
  typeof value === "string" && value.trim() !== "";

/**
 * An environment name as the platform stores it: lowercase, `[a-z0-9_-]`
 * only, at most 32 characters. Anything else collapses to a dash, and an
 * empty result is the default environment.
 */
export function sanitizeEnvironment(name: string): string {
  const cleaned = name
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9_-]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, ENVIRONMENT_MAX_LENGTH)
    .replace(/-+$/g, "");
  return cleaned === "" ? DEFAULT_ENVIRONMENT : cleaned;
}

/**
 * The environment an agent registers under: the explicit option, then
 * `LANGWATCH_AGENT_ENVIRONMENT`, `APP_ENV`, `ENVIRONMENT`, `NODE_ENV`, else
 * `development`.
 */
export function resolveEnvironment({
  explicit,
  env = process.env,
}: {
  explicit?: string;
  env?: NodeJS.ProcessEnv;
}): string {
  if (isSet(explicit)) return sanitizeEnvironment(explicit);
  for (const name of ENVIRONMENT_VARIABLES) {
    const value = env[name];
    if (isSet(value)) return sanitizeEnvironment(value);
  }
  return DEFAULT_ENVIRONMENT;
}

const isTruthy = (value: string | undefined): boolean => {
  if (!isSet(value)) return false;
  const lowered = value.trim().toLowerCase();
  return lowered !== "0" && lowered !== "false" && lowered !== "no" && lowered !== "off";
};

/**
 * Whether the agent connects at all. `LANGWATCH_AGENT_CONNECT=0` (or false)
 * always disables it; the explicit option wins next; otherwise the connection
 * is on, except when `CI` is truthy.
 */
export function resolveEnabled({
  explicit,
  env = process.env,
}: {
  explicit?: boolean;
  env?: NodeJS.ProcessEnv;
}): boolean {
  const flag = env.LANGWATCH_AGENT_CONNECT;
  if (isSet(flag) && !isTruthy(flag)) return false;
  if (explicit !== undefined) return explicit;
  return !isTruthy(env.CI);
}

/** The instance label: the option, then `LANGWATCH_AGENT_INSTANCE_LABEL`. */
export function resolveInstanceLabel({
  explicit,
  env = process.env,
}: {
  explicit?: string;
  env?: NodeJS.ProcessEnv;
}): string | undefined {
  if (isSet(explicit)) return explicit.trim();
  const fromEnv = env.LANGWATCH_AGENT_INSTANCE_LABEL;
  return isSet(fromEnv) ? fromEnv.trim() : undefined;
}

/** The two reads of the machine, replaceable so a test can make them fail. */
export interface MachineReader {
  hostname: () => string;
  userInfo: () => { username: string };
}

const HOST_LABEL_MAX_LENGTH = 24;

/**
 * A short label for this machine: lowercase, `[a-z0-9-]`, 24 characters.
 *
 * The platform scopes a development agent connected with a project key to
 * this label, and the Python SDK sends the same shape, so one machine reads
 * the same whichever SDK connected it.
 */
export function hostLabel(hostname: string): string {
  return hostname
    .toLowerCase()
    .replace(/\.(local|lan|home|localdomain)$/i, "")
    .replace(/[^a-z0-9-]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, HOST_LABEL_MAX_LENGTH);
}

const readHostname = (machine: MachineReader): string => {
  try {
    return hostLabel(machine.hostname());
  } catch {
    return "";
  }
};

const readUsername = (machine: MachineReader): string => {
  try {
    return machine.userInfo().username;
  } catch {
    return "";
  }
};

/** The identity one process announces in `register`, built once. */
export function buildInstance({
  label,
  machine = os,
}: {
  label?: string;
  machine?: MachineReader;
}): RegisterInstance {
  return {
    id: `inst_${randomUUID().replace(/-/g, "")}`,
    hostname: readHostname(machine),
    username: readUsername(machine),
    pid: process.pid,
    startedAt: new Date().toISOString(),
    ...(label ? { label } : {}),
    inFlightCallIds: [],
  };
}

/** The SDK block of `register`. */
export const SDK_IDENTITY: RegisterSdk = {
  name: "langwatch-typescript",
  version: LANGWATCH_SDK_VERSION,
  language: "typescript",
};

export const USER_AGENT = `langwatch-typescript/${LANGWATCH_SDK_VERSION}`;

export const CONNECT_PATH = "/api/agents/connect";

/**
 * The socket URL for an endpoint: `https://app.langwatch.ai` becomes
 * `wss://app.langwatch.ai/api/agents/connect`, `http://localhost:5560`
 * becomes `ws://localhost:5560/api/agents/connect`.
 */
export function resolveConnectUrl(endpoint?: string | null): string {
  const base = resolveEndpoint(endpoint);
  const socketBase = base.replace(/^http(s?):\/\//i, (_match, secure: string) =>
    secure ? "wss://" : "ws://",
  );
  return `${socketBase}${CONNECT_PATH}`;
}

/** The headers the socket opens with. */
export function buildConnectHeaders({
  apiKey,
  projectId,
}: {
  apiKey: string;
  projectId?: string;
}): Record<string, string> {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${apiKey}`,
    "User-Agent": USER_AGENT,
  };
  if (isSet(projectId)) headers["X-Project-Id"] = projectId;
  return headers;
}

File: sdks/python/src/langwatch/agent/identity.py
"""Environment and instance identity of a connected agent process.

The platform keys an agent row by name and environment, and lists the
instances of a row by hostname, username, pid and label. Every lookup here
is defensive: a sandbox without a hostname or a container whose uid has no
passwd entry gives an empty string, never an exception.
"""

from __future__ import annotations

import getpass
import os
import re
import socket
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone

from .protocol import InstanceInfo

DEFAULT_ENVIRONMENT = "development"

ENVIRONMENT_VARIABLES = (
    "LANGWATCH_AGENT_ENVIRONMENT",
    "APP_ENV",
    "ENVIRONMENT",
    "NODE_ENV",
)

CONNECT_VARIABLE = "LANGWATCH_AGENT_CONNECT"
INSTANCE_LABEL_VARIABLE = "LANGWATCH_AGENT_INSTANCE_LABEL"

MAX_ENVIRONMENT_LENGTH = 64
MAX_LABEL_LENGTH = 64
MAX_HOST_LABEL_LENGTH = 24

_FALSE_VALUES = frozenset({"0", "false", "no", "off"})
_TRUE_VALUES = frozenset({"1", "true", "yes", "on"})

_HOST_SUFFIX_RE = re.compile(r"\.(local|lan|home|localdomain)$", re.IGNORECASE)
_NOT_HOST_CHARS_RE = re.compile(r"[^a-z0-9-]+")
_NOT_ENVIRONMENT_CHARS_RE = re.compile(r"[^a-z0-9_.-]+")
_EDGE_DASHES_RE = re.compile(r"^-+|-+$")


def sanitize_environment(value: str | None) -> str:
    """Lowercase, replace symbols and spaces with dashes, cap the length.

    An empty result falls back to `development`.
    """
    if value is None:
        return DEFAULT_ENVIRONMENT
    cleaned = _NOT_ENVIRONMENT_CHARS_RE.sub("-", value.strip().lower())
    cleaned = _EDGE_DASHES_RE.sub("", cleaned)[:MAX_ENVIRONMENT_LENGTH]
    cleaned = _EDGE_DASHES_RE.sub("", cleaned)
    return cleaned or DEFAULT_ENVIRONMENT


def resolve_environment(explicit: str | None = None) -> str:
    """The explicit argument, then the environment variables in order."""
    if explicit is not None and explicit.strip():
        return sanitize_environment(explicit)
    for variable in ENVIRONMENT_VARIABLES:
        value = os.environ.get(variable)
        if value is not None and value.strip():
            return sanitize_environment(value)
    return DEFAULT_ENVIRONMENT


def _flag_from_environment(variable: str) -> bool | None:
    raw = os.environ.get(variable)
    if raw is None:
        return None
    value = raw.strip().lower()
    if value in _FALSE_VALUES:
        return False
    if value in _TRUE_VALUES:
        return True
    return None


def ci_is_truthy() -> bool:
    raw = os.environ.get("CI")
    if raw is None:
        return False
    return raw.strip().lower() not in ("", *_FALSE_VALUES)


def resolve_enabled(explicit: bool | None = None) -> bool:
    """`LANGWATCH_AGENT_CONNECT=0` always disables.

    Otherwise the explicit argument decides; with no argument the connection
    is on, except when `CI` is truthy.
    """
    from_variable = _flag_from_environment(CONNECT_VARIABLE)
    if from_variable is False:
        return False
    if explicit is not None:
        return explicit
    if from_variable is True:
        return True
    return not ci_is_truthy()


def sanitize_label(value: str | None) -> str | None:
    if value is None:
        return None
    cleaned = " ".join(value.split())[:MAX_LABEL_LENGTH].strip()
    return cleaned or None


def resolve_instance_label(explicit: str | None = None) -> str | None:
    """The explicit argument, then `LANGWATCH_AGENT_INSTANCE_LABEL`."""
    label = sanitize_label(explicit)
    if label is not None:
        return label
    return sanitize_label(os.environ.get(INSTANCE_LABEL_VARIABLE))


def machine_hostname() -> str:
    try:
        return socket.gethostname() or ""
    except Exception:
        return ""


def machine_username() -> str:
    try:
        return getpass.getuser() or ""
    except Exception:
        return ""


def host_label(hostname: str | None = None) -> str:
    """A short label for this machine: lowercase, `[a-z0-9-]`, 24 characters.

    The platform scopes a development agent connected with a project key to
    this label, so it has the same shape as the label the CLI attaches to the
    keys it mints.
    """
    raw = (hostname if hostname is not None else machine_hostname()).lower()
    cleaned = _HOST_SUFFIX_RE.sub("", raw)
    cleaned = _NOT_HOST_CHARS_RE.sub("-", cleaned)
    cleaned = _EDGE_DASHES_RE.sub("", cleaned)
    return cleaned[:MAX_HOST_LABEL_LENGTH]


def new_instance_id() -> str:
    return uuid.uuid4().hex


@dataclass
class InstanceIdentity:
    """What the platform shows for one connected process."""

    id: str = field(default_factory=new_instance_id)
    hostname: str = field(default_factory=machine_hostname)
    username: str = field(default_factory=machine_username)
    pid: int = field(default_factory=os.getpid)
    started_at: str = field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )
    label: str | None = None

    def to_frame(self, *, in_flight_call_ids: list[str]) -> InstanceInfo:
        info: InstanceInfo = {
            "id": self.id,
            "hostname": host_label(self.hostname),
            "username": self.username,
            "pid": self.pid,
            "startedAt": self.started_at,
            "inFlightCallIds": list(in_flight_call_ids),
        }
        if self.label:
            info["label"] = self.label
        return info
